Add property checker/metadata system and C++ helpers

garyo opened the pull request:
* `include/ofx-props.yml` is the property reference/metadata
* `scripts/gen-props.py` is the checker and source generator

All props are included in the YML definition, and the script generates C++ files with all the property metadata and suite assignments. The C++ files aren't yet checked in, but you can build them easily: `python scripts/gen-props.py -v` and then look in `include/gen_*.hxx`.

garyo committed:
Add property checker/metadata system

* include/ofx-props.yml is the property reference/metadata
* scripts/gen-props.py is the checker and source generator

All props are included in the YML definition, and the script generates
C++ files with all the property metadata and suite assignments.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Props metadata: fixes, namespacing, sorting

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add all actions with their property sets, clean up misc list

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Generate list of all actions, & fix action names

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Commit generated source files, with proper headers

Also include them into prop tester plugin, to test whether they
compile properly. TODO: prop tester should use the metadata to do
better tests.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Fix CI build: ofxPropsBySet needs <array>

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

abizeauMaxon commented on include/ofxPropsMetadata.h:
```diff
@@ -0,0 +1,224 @@
+  std::vector<const char *> values; // for enums
+};
+
+const std::vector<struct PropsMetadata> props_metadata {
```
Since this it auto generated, you could consider `std::array<struct PropsMetadata, N=FeedByScrip>`.

If this header is intended to be use more C++, you should consider having this `std::array` as constexpr.

maybe:

```
#if defined(__cplusplus)
#define CONST_SPECIFIER constexpr
#else
#define CONST_SPECIFIER const
#endif
```

garyo replied:
Thanks @abizeauMaxon -- totally agreed. As per discussion at the TSC today, I've moved these into `Support/include` and they are 100% C++. I used fixed-size `std::array`s where possible, as well as some other updates and cleanups. More feedback will be helpful; let me know if you try it out.

abizeauMaxon commented on include/ofxPropsMetadata.h:
```diff
@@ -0,0 +1,224 @@
+  int dimension;
+  Writable writable;
+  bool host_optional;
+  std::vector<const char *> values; // for enums
```
You should be consistant, if you use `std::string` for name, you should also use `std::vector<std::string>`

garyo replied:
I've tried now to move to `char *`, to avoid allocations where possible during startup. I should move to `std::string_view` in some places. I'll do that.

abizeauMaxon commented on include/ofxPropsBySet.h:
```diff
@@ -0,0 +1,733 @@
+
+namespace OpenFX {
+// Properties for property sets
+const std::map<const char *, std::vector<const char *>> prop_sets {
```
again, here you should consider consistency, pick `std::string` or `const char*`

abizeauMaxon commented on include/ofx-props.yml:
```diff
@@ -0,0 +1,1508 @@
+########################################################################
```
Is this auto generated? If so maybe add it into the header

garyo replied:
`ofx-props.yml` is hand-created -- it's now the source of truth for all info about properties and property sets. That's the lion's share of the work in this PR (and probably the source of remaining bugs, where I may have gotten some stuff wrong). None of this info was machine-readable before, so I extracted it from the docs and header file comments.

abizeauMaxon reviewed (commented):
That super! I see a lot of good thing here! Will be helpfull with time.

garyo committed:
ofx-props.yml: Remove writable defs, not the right place for that

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Props metadata: use strings, not C #define names, for prop names

The strings are the true spec, not the #define names. This also adds a
set of static_assert tests to ensure the #define names match their
strings. But doing this brought up that there are some mismatches
between the strings and their #defines. Those will be accounted for
in a following commit.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Account for some prop names that don't match their C #defines

See `get_cname` and `find_stringname` in gen-props.py, and the new
`cname` member of the props metadata.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Make all generated objects static inline, clean up

Replaced a vector with a std::array for reduced memory & startup time.
Also add static_assert checks for all action names vs. their #defines.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add "introduced" for new props in 1.5

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add support for host/plugin write status in props_by_set

Also separate out Actions from prop sets in ofx-props.yml, for
cleanliness.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Use string_view instead of strings or char* in generated files

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
ofx-props.yml: Add GetOutputColourspace action

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add TestProps using new props metadata, WIP

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Support: Props accessor using new metadata

Update testProperties sample to use it (WIP)

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
WIP: start of openfx-cpp C++ bindings

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Clean up generated ofxPropsMetadata

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Improve OfxRect converters, fix #includes and macros

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Fix prop defs & add ofxClip.h with more clip/image accessors

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Fix prop types for RoD and RoI (should be double)

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add property setters for OfxRect and OfxPoint

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add suite container, std exceptions, more accessors, fixes

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Generate property & prop sets reference doc from ofx-props.yml

- New script gen-props-doc.py to generate RST doc for all props and
  prop sets
  - Generated docs are checked in, just like the generated props
    headers.
- Use it in Documentation/build.sh
- Fix some other doc I noticed along the way

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Try to fix CI build by allowing older conan version (>=2.0.16)

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Rebase to main & include new prop PropNoSpatialAwareness

- Added to ofx-props.yml
- Ran gen-props.py script to update property set headers

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add links from Properties Reference to the structured docs

Also cleaned up a few doc typos I noticed.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add ofxSpan.h for spans, using tcb-span or std::span

Switch to using spans in a few places where it makes sense, mostly
for property type arrays.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Add system for type-safe host-specific properties

Hosts can set up properties in their own namespaces and give plugins
type-safe access just like built-in properties.

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Checkpoint: template for host-specific properties

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Remove unneeded `id` PropDef struct member & clean up

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Test compliance for all props, & add propset accesssors

To be extended to all standard prop sets

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>

garyo committed:
Work on testProperties: now tests compliance for many propsets

Signed-off-by: Gary Oberbrunner <garyo@darkstarsystems.com>
